common/testing.cpp

```diff
@@ -89,7 +89,14 @@ const string HttpFileServer::serve_address_ {"http://127.0.0.1:53272"};
 HttpFileServer::HttpFileServer(const string &dir) :
 	dir_(dir),
 	server_(http::ServerConfig {}, loop_) {
-	thread_ = thread([this]() {
+	// The reason we need this synchronization is because of the thread sanitizer and
+	// logging. AsyncServeUrl uses the logger internally, and the log level is also set by
+	// certain tests. Since these things happen in two separate threads, we need to make sure
+	// that AsyncServeUrl has returned before we leave this function.
+	promise<bool> running;
+	auto maybe_running = running.get_future();
+
+	thread_ = thread([this, &running]() {
```

Comment:
Why do we even need a thread here? It should be possible to use the same loop for `AsyncServeUrl()` and for the clients, just like we do in the other HTTP tests not using this new class, right?

Reply:
Yes, it's possible. However, it's very tricky in the case where I'm using it, because we are calling the binary's `Main()` function. Which is long before an `EventLoop` has been created. And fishing it out during the running of `Main()` would also be tricky, and would need test callback of some sort. So the thread just makes it easier with a fully standalone server. It can almost be considered a separate process, semantically speaking.